Direct ARGB encoding, texture and non-texture (RGB)

cmlin2 opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
<!--- Describe your changes in detail. -->
<!--- If this change includes UI elements, please include screenshots. -->
Allow for direct ARGB (RGB) non-texture and texture encoding, bypassing unneeded NV12 conversion

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open GitHub Issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->
QSV from ICL+ allows for direct ARGB HW conversion so it improves CPU performance to use it (as unneeded RGB->NV12 conversion is done on CPU)

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->
Tested on CFL machine with NVIDIA card and TGL machine on Windows, making RGB (non-texture and texture) selections and using Windows->Graphics Settings to change if the renderer and encoder are on the same GPU by selecting Power Saving or Let Windows Decide/High Performance. Results on both CFL and TGL (so machines that are ICL- and ICL+) show CPU improvement and demonstrate functionality with and without discrete graphics. 

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [X] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [X] My code is not on the master branch.
- [X] The code has been tested.
- [X] All commit messages are properly formatted and commits squashed where appropriate.
- [X] I have included updates to all appropriate documentation.

RytoEX commented on libobs/obs-video.c:
```diff
@@ -322,6 +382,12 @@ static void render_convert_texture(struct obs_core_video *video,
+
+		// Added for tracing direct ARGB encoding
+		blog(LOG_INFO,
+		     "=== [obs-video] render_convert_texture, from %p to %p",
```
This appears to be called on every single frame, which creates a huge amount of log spam.  This should either be LOG_DEBUG, or ifdef'd out when not needed, or not included at all in a final merge.

RytoEX commented on libobs/obs.c:
```diff
@@ -357,6 +377,7 @@ static int obs_init_graphics(struct obs_video_info *ovi)
 		success = false;
 
 	gs_leave_context();
+
```
Don't introduce unnecessary whitespace changes.

RytoEX reviewed (changes requested):
This briefly broke all non-Windows builds.  Testing on non-Windows platforms seems appropriate, given how much core libobs code is being touched.
https://github.com/obsproject/obs-studio/actions/runs/959307724

Please follow our [Commit Guidelines](https://github.com/obsproject/obs-studio/blob/8ec73aea860d6ebb4b97d47201cf2784da8a9da2/CONTRIBUTING.rst#commit-guidelines) for your commit message.  Please don't include fixup commits or merge commits in your commit history for PRs.  Please appropriately squash commits.

> QSV from ICL+ allows for direct ARGB conversion so it improves performance to use it

> Tested on CFL machine with NVIDIA card on Windows

Is the fact that an NVIDIA dGPU was used relevant for reproducing your test results?  Does this work on a system with an AMD dGPU, or without any dGPU?

Am I mistaken, or is CFL (Coffee Lake?) older than ICL+ (Ice Lake and newer?) and thus wouldn't benefit from these optimizations?

> Tested on CFL machine with NVIDIA card on Windows, making RGB (non-texture and texture) selections and using Graphics Settings to change if the renderer and encoder are on the same GPU

What selections are you making where?  What Graphics Settings -  Windows, OBS, NVIDIA Control Panel, Intel Graphics Command Center?

cmlin2 commented:
I will make a new pull request that addresses all your comments. I'm having some problems squashing with a merge commit so it's easier for me just to start over and have one clean commit.

RytoEX commented:
> 
> 
> I will make a new pull request that addresses all your comments. I'm having some problems squashing with a merge commit so it's easier for me just to start over and have one clean commit.

Creating a new pull request forces us to refer back to this one for comment and review history and ensure that the review comments have actually been addressed, which is not ideal.

In your local Git repo, if you have your fork (cmlins2/obs-studio) set as a remote named "origin" and obsproject/obs-studio set as a separate remote named "upstream" (you can check your remotes by executing `git remote -v`), then the following should get you to a cleaner state:

```
git checkout ARGBdirect
git rebase -i HEAD~3
```

This will open your default text editor, which will ask you to select actions for rebasing.  You'll have a list like this:
```
pick c53a98568 Direct ARGB encoding, texture and non-texture (RGB)
pick 9dd818845 Direct ARGB encoding, texture and non-texture (RGB)
pick c301ba216 Remove spurious include of windows.h.
```

Leave the first line alone.  For the second and third lines, change "pick" to "f".  Save the file.  Close the editor.  Git should finish rebasing.  Then you can force push to your remote branch, https://github.com/cmlin2/obs-studio/tree/ARGBdirect.

After that, you can amend the commit to update its contents and fix the commit wording and force push those updates.

cmlin2 commented:
Thanks @RytoEX for the tips. I was squashing okay, but didn't realize I had to force push due to that. Hopefully you can now see the requested changes.

[Enabling direct ARGB encoding in OBS.pptx](https://github.com/obsproject/obs-studio/files/6696292/Enabling.direct.ARGB.encoding.in.OBS.pptx)

RytoEX commented on plugins/obs-qsv11/obs-qsv11.c:
```diff
@@ -979,6 +1002,9 @@ static bool obs_qsv_encode_tex(void *data, uint32_t handle, int64_t pts,
 {
 	struct obs_qsv *obsqsv = data;
 
+	// Added for debug of adding direct ARGB encoding
+	blog(LOG_INFO, "=== [qsv] receive texture, handle %p", handle);
+
```
Change to LOG_DEBUG, or use an ifdef, or remove this.

If keeping this, please note that this spams the log on every frame for any GPU encoder, such as NVENC, with messages about QSV.

RytoEX reviewed (changes requested):
This PR spams the OBS log even in release configurations of OBS.  The log is spammed on every frame with multiple debug messages, even when encoding with NVENC.

This PR also breaks NVENC output with RGB color format on Windows on my GTX 1060.  I get the following in my OBS log:
```
device_copy_texture (D3D11): Source and destination formats do not match
```
And resulting video file is blank.

This PR also breaks QSV output with RGB color format on Windows with my i7-7700HQ (HD Graphics 630) regardless if OBS is running on the GTX 1060 or on the HD Graphics 630.  I get the same error message as above and the resulting video file is blank.

These configurations work fine on OBS Studio 27.0.1.

cmlin2 commented:
I have updated the LOG_INFO messages with LOG_DEBUG. Will root-cause and resolve the blank video issues.

cmlin2 commented:
I have updated the QSV direct ARGB code to fix the bugs causing blank videos and errors.

RytoEX commented:
The PR description says:
> QSV from ICL+ allows for direct ARGB HW conversion so it improves CPU performance to use it (as unneeded RGB->NV12 conversion is done on CPU)

The [code in this PR](https://github.com/obsproject/obs-studio/pull/4931/files#diff-148b1d740b61232c8f1c97089d6d00e44f4b95c00b989c960d9373168f08fdd1R809-R815):
```
	// Platforms >= KBL will have direct ARGB conversion support as well as NV12
	if (qsv_platform >= QSV_CPU_PLATFORM_KBL) { // QSV_CPU_PLATFORM_ICL
		return (format == VIDEO_FORMAT_NV12) ||
		       (format == VIDEO_FORMAT_RGBA);
	} else {
		return format = VIDEO_FORMAT_NV12;
	}
```


Is this meant to work on Kaby Lake as well? Or was this in place for some kind of testing? I'd like to be clear on the hardware restrictions so that we can properly identify test platforms. For example, I work on a Kaby Lake system, so I'd like to know what to expect on my system.

cmlin2 commented:
Here are further details for AVC ARGB support courtesy of @YaoxinShi:
- From API/application perspective ARGB input is supported for all Gens started from Gen9, both VME and VDEnc.
- If there is no native ARGB support – use EU- or SFC-based Color-Space-Conversion embedded to Media Driver directly (all CSC logic is hidden from user)
- For AVC VME – NV12 is the only one native color format -> CSC from RGB to NV12 has been used for all Gens
- For AVC VDEnc:
   > For DG2+ RGB input become a native (both Linear and Tiled) – put directly to VDEnc
   > For ICL+ - only Linear RGB is a native, for Tiled there is also CSC call
   > For SKL/KBL - RGB is not native and CSC is used

openglfreak commented:
@cmlin2
I don't think adding a new flag for each supported texture format is the best way to do this - imo it's better to do it the way I did in #5706: Add a callback to the encoder which returns whether a format is supported. My PR also includes QSV support, taken from this PR, so I think it fully supersedes your this one.

ValZapod commented:
>as unneeded RGB->NV12 conversion is done on CPU

Which you should never do. YCbCr 8 bit even 444 is not enough to preserve RGB 8 bit, and NV12 is 8 bit 420. Use at least P010 (YCbCr 10 bit 420), that is enough to get round trip.

RytoEX commented:
After further discussion with Intel, we've decided to let PR #5706 supersede this one.
